## Part 1: Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [2]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [3]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]
print(y_df.dtypes)

Attrition     object
Department    object
dtype: object


In [4]:
# Create a list of at least 10 column names to use as X data
# columns = ["Age", "BusinessTravel", "DistanceFromHome", "Education", "HourlyRate", "JobSatisfaction", "PerformanceRating", "EnvironmentSatisfaction", "WorkLifeBalance", "YearsInCurrentRole"]
columns = ["Age", "BusinessTravel", "DistanceFromHome", "HourlyRate", "JobSatisfaction", "PerformanceRating", "WorkLifeBalance", "YearsInCurrentRole"]

# Create X_df using your selected columns
X_df = attrition_df[columns]

# Because BusinessTravel is a categorical variable, we need to convert it to a numerical value.
# Convert the BusinessTravel column to a numerical value
X_df = pd.get_dummies(X_df, columns=["BusinessTravel"], dtype=int)

# Show the data types for each column including the new dummy columns
print(X_df.dtypes)

Age                                 int64
DistanceFromHome                    int64
HourlyRate                          int64
JobSatisfaction                     int64
PerformanceRating                   int64
WorkLifeBalance                     int64
YearsInCurrentRole                  int64
BusinessTravel_Non-Travel           int32
BusinessTravel_Travel_Frequently    int32
BusinessTravel_Travel_Rarely        int32
dtype: object


In [5]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=78)

In [6]:
# Convert your X data to numeric data types however you see fit
X_train = X_train.astype('int64')
X_test = X_test.astype('int64')

###### Add new code cells as necessary
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(1102, 10) (368, 10) (1102, 2) (368, 2)


In [7]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the StandardScaler to the training data
X_scaler = scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [8]:
# Create a OneHotEncoder for the Department column
from sklearn.preprocessing import OneHotEncoder
enc_Department = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
enc_Department.fit(y_train['Department'].values.reshape(-1, 1))

# Create two new variables by applying the encoder to the training and testing data
y_train_encoded_Department = enc_Department.transform(y_train['Department'].values.reshape(-1, 1))
y_test_encoded_Department = enc_Department.transform(y_test['Department'].values.reshape(-1, 1))

# print first 5 rows of y_test_encoded_Department
y_test_encoded_Department[:5]

array([[0., 1., 0.],
       [0., 1., 0.],
       [0., 1., 0.],
       [0., 1., 0.],
       [0., 1., 0.]])

In [9]:
# Create a OneHotEncoder for the Attrition column
enc_Attrition = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
enc_Attrition.fit(y_train['Attrition'].values.reshape(-1, 1))

# Create two new variables by applying the encoder to the training and testing data
y_train_encoded_Attrition = enc_Attrition.transform(y_train['Attrition'].values.reshape(-1, 1))
y_test_encoded_Attrition = enc_Attrition.transform(y_test['Attrition'].values.reshape(-1, 1))

# print first 5 rows of y_test_encoded_Attrition
y_test_encoded_Attrition[:5]

array([[1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [0., 1.]])

## Create, Compile, and Train the Model

In [10]:
# Find the number of columns in the X training data
input_features = X_train_scaled.shape[1]

# Create the input layer
inputs = layers.Input(shape=(input_features))   # Create input layer

# Create at least two shared layer
shared_1 = layers.Dense(64, activation='relu', name='shared1')(inputs)
shared_2 = layers.Dense(128, activation='relu', name='shared2')(shared_1)

In [11]:
# Create a branch for Department with a hidden layer and an output layer
# Create the hidden layer
hidden_1 = layers.Dense(32, activation='relu', name='department_hidden')(shared_2)

# Create the output layer
output_1 = layers.Dense(3, activation='softmax', name='department_output')(hidden_1)

In [12]:
# Create a branch for Attrition with a hidden layer and an output layer
# Create the hidden layer
hidden_2 = layers.Dense(32, activation='relu', name='attrition_hidden')(shared_2)

# Create the output layer
output_2 = layers.Dense(2, activation='sigmoid', name='attrition_output')(hidden_2)

In [13]:
# Create the model
model = Model(inputs=inputs, outputs=[output_1, output_2])

# Compile the model
model.compile(optimizer='adam',
              loss={'department_output': 'categorical_crossentropy', 'attrition_output': 'binary_crossentropy'},
              metrics={'department_output': 'accuracy', 'attrition_output': 'accuracy'})

# Summarize the model
model.summary()


Model: "model"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_1 (InputLayer)        [(None, 10)]                 0         []                            
                                                                                                  
 shared1 (Dense)             (None, 64)                   704       ['input_1[0][0]']             
                                                                                                  
 shared2 (Dense)             (None, 128)                  8320      ['shared1[0][0]']             
                                                                                                  
 department_hidden (Dense)   (None, 32)                   4128      ['shared2[0][0]']             
                                                                                             

In [14]:
# Train the model
model.fit(X_train_scaled, {'department_output': y_train_encoded_Department, 'attrition_output': y_train_encoded_Attrition}, epochs=100, shuffle=True, verbose=2)

Epoch 1/100


35/35 - 2s - loss: 1.3380 - department_output_loss: 0.8250 - attrition_output_loss: 0.5130 - department_output_accuracy: 0.6642 - attrition_output_accuracy: 0.8221 - 2s/epoch - 71ms/step
Epoch 2/100
35/35 - 0s - loss: 1.1988 - department_output_loss: 0.7567 - attrition_output_loss: 0.4421 - department_output_accuracy: 0.6670 - attrition_output_accuracy: 0.8276 - 54ms/epoch - 2ms/step
Epoch 3/100
35/35 - 0s - loss: 1.1783 - department_output_loss: 0.7572 - attrition_output_loss: 0.4212 - department_output_accuracy: 0.6670 - attrition_output_accuracy: 0.8303 - 61ms/epoch - 2ms/step
Epoch 4/100
35/35 - 0s - loss: 1.1499 - department_output_loss: 0.7405 - attrition_output_loss: 0.4094 - department_output_accuracy: 0.6688 - attrition_output_accuracy: 0.8385 - 47ms/epoch - 1ms/step
Epoch 5/100
35/35 - 0s - loss: 1.1374 - department_output_loss: 0.7335 - attrition_output_loss: 0.4039 - department_output_accuracy: 0.6688 - attrition_output_accuracy: 0.8430 - 47ms/epoch - 1ms/step

In [15]:
# Evaluate the model with the testing data
results = model.evaluate(X_test_scaled, {'department_output': y_test_encoded_Department, 'attrition_output': y_test_encoded_Attrition}, verbose=2)
print(results)

12/12 - 0s - loss: 3.0724 - department_output_loss: 2.2202 - attrition_output_loss: 0.8522 - department_output_accuracy: 0.5109 - attrition_output_accuracy: 0.8071 - 337ms/epoch - 28ms/step
[3.072410821914673, 2.220249652862549, 0.8521612286567688, 0.510869562625885, 0.8070651888847351]


In [16]:
# Print the accuracy for both department and attrition
print(f"Department Accuracy: {results[3]}")
print(f"Attrition Accuracy: {results[4]}")


Department Accuracy: 0.510869562625885
Attrition Accuracy: 0.8070651888847351


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. I think **accuracy** can be used in this model to evaluate the model's performance. The model has an accuracy of 0.80 for the department prediction and 0.51 for the attrition prediction. This means that the model is able to predict the department of an employee with 83% accuracy and whether an employee will leave the company with 85% accuracy. This is a good result, but there is still room for improvement. The model can be further optimized by tuning the hyperparameters, increasing the number of epochs, or adding more layers to the neural network. Additionally, more data can be collected to improve the model's performance. Overall, the model is a good starting point for HR to predict employee attrition and department fit.

This type of accuracy may not be very reliable if there were cost sensitive issues. For example, if the cost of false positives is high, then the model may not be very useful. In this case, the model may need to be further optimized to reduce the number of false positives. However, if the cost of false negatives is high, then the model may be useful as is. It is important to consider the cost of false positives and false negatives when evaluating the model's performance. While predicting attrition for any company is important, but cannot be compared with outcome of for example bad fraud detection or healthcare issues.

2. I chose `relu` for shared layer because it is a good activation function for hidden layers in neural networks. It is able to handle the vanishing gradient problem, which can occur when training deep neural networks. I chose `sigmoid` for the attrition output layer because it is a good activation function for binary classification problems. It squashes the output to a range between 0 and 1, which is useful for predicting probabilities. I chose `softmax` for the department output layer because it is a good activation function for multi-class classification problems. It squashes the output to a range between 0 and 1, which is useful for predicting probabilities. Additionally, it ensures that the sum of the output values is equal to 1, which is useful for interpreting the results.

3. A model might be improved by increasing the number of epochs, adding more layers to the neural network, or tuning the hyperparameters. Additionally, more data can be collected to improve the model's performance. Overall, the model is a good starting point for HR to predict employee attrition and department fit. Also by chossing different data column to understand which data column might be produce most accurate results. 